### ISSUE TITLE
Question about having better performance

### ISSUE THREAD

**Issue by ghZHM** [ultralytics/yolov5#6994]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello, I follow the instruction in the document to train my custom data.
Here is the label figure. 
![labels](https://user-images.githubusercontent.com/98072324/158496288-085b0591-903d-4650-9f9b-9a53e5e48f24.jpg)

Since I only want my model to detect these 4 classes,I just ignore the annotation of other classes in the original BDD100k dataset.
    As you recommended in the document, I start to train use following command
    train.py --img 640 --batch 48 --epochs 300 --data bdd.yaml --weights yolov5s.pt
    And here is the result
![results](https://user-images.githubusercontent.com/98072324/158496326-3c7e8f81-8416-476c-a190-0132b9b49c4d.png)
![log](https://user-images.githubusercontent.com/98072324/158496410-92fb799c-92be-4fd1-a13c-a78b5cd864e9.png)

    As you mentioned both in the document and issues raised by others, train until overfitting occur
    I found that cls_loss in val begin to increase from the 140 epoch, but other two loss just keep decreasing.
    And obj_loss slightly went up at about 200 epoch.
    I searched in the issues, and found all these three loss component need to be overfitting.
    I also learn that if they can overfit in a similar time, although it is hard, it maybe better for the model
    Besides, the best weight was obtained in the 144 epoch.
    So could you please give me some advice on how to train a better model.
    Really appreciate for your help.

### Additional

Here is some charts in the train folder

![confusion_matrix](https://user-images.githubusercontent.com/98072324/158496534-3cae3753-c2ac-4be9-aa40-2909add006fb.png)
![PR_curve](https://user-images.githubusercontent.com/98072324/158496560-f1fb8930-b531-4f62-8bc9-67e469d2c658.png)
![R_curve](https://user-images.githubusercontent.com/98072324/158496565-93e54e18-b70b-4820-95ed-80f33af82fb4.png)
![F1_curve](https://user-images.githubusercontent.com/98072324/158496606-b333e9e2-8233-48a1-b552-aa41e2a8e1ce.png)
![P_curve](https://user-images.githubusercontent.com/98072324/158496607-f34ae2b6-b76b-4f3e-a6af-2780940fd47e.png)
![train_batch1](https://user-images.githubusercontent.com/98072324/158496608-187c01b0-2087-44aa-a5cd-6754c5acb7c7.jpg)
![train_batch2](https://user-images.githubusercontent.com/98072324/158496609-b0bf9454-8df1-4fa0-ae4d-bfebea0e2bbc.jpg)

**Comment by glenn-jocher** [comment]
@ghZHM 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by ghZHM** [comment]
Thank you for your reply, I will train with default setting now. Besides, I have another question about my custom training data.
It used to be the BDD100K dataset.
![instance](https://user-images.githubusercontent.com/98072324/158720742-ebcbca9d-10f2-4839-a7da-af40a1995f8d.png)

Since I only need the model to detect pedestrian, car, bus and truck, I didn't keep the annotation of other classes in labels/.txt annotation file. Is it correct? Or I should keep the annotations to avoid misleading the model

**Comment by glenn-jocher** [comment]
@ghZHM if you'd like to use a subset of your dataset classes you can either customize labels to suit your needs and supply full labels and filter here:
https://github.com/ultralytics/yolov5/blob/3f634d43c8ecea14aa9037e2fd28ded0433d491d/utils/datasets.py#L451

BTW BDD100K is a popular dataset that we don't have a yaml for. Please consider submitting a PR for a BDD100K.yaml file to help future users!

**Comment by ghZHM** [comment]
I would like to help for the BDD100K.yaml. But I could not deal with the download script part in yaml, the dataset I have was downloaded into a hard disk drive in US and sent me by mail. And the other items like where you save the images or how many class in it , I think maybe leave them to future users to fill it since it is really simple.
